compiler/src/iree/compiler/DispatchCreation/SetEncoding.cpp

```diff
@@ -227,17 +229,45 @@ class SetContractionOpEncoding final
     SmallVector<AffineMap> maps = linalgOp.getIndexingMapsArray();
 
     auto opType = IREE::Encoding::EncodingOpType::matmul;
-    auto setEncodingWrapper = [&](Value src, int64_t operandIndex) -> Value {
-      auto encoding =
-          EncodingAttr::get(linalgOp.getContext(), operandIndex, opType,
-                            elemTypes, maps, iterationSizes);
+    auto setEncodingWrapper = [&](Value src,
+                                  int64_t operandIndex) -> FailureOr<Value> {
+      MLIRContext *ctx = linalgOp.getContext();
+      Attribute encoding;
+      switch (encodingOption) {
+      case EncodingOptions::Default: {
+        encoding = EncodingAttr::get(ctx, operandIndex, opType, elemTypes, maps,
+                                     iterationSizes);
+        break;
+      }
+      case EncodingOptions::MatmulK: {
+        SmallVector<int32_t> kDims;
+        AffineMap indexingMap = maps[operandIndex];
+        auto cDims = linalg::inferContractionDims(linalgOp);
+        for (auto k : cDims->k) {
+          std::optional<unsigned> dimIdx =
+              indexingMap.getResultPosition(rewriter.getAffineDimExpr(k));
+          if (!dimIdx) {
+            continue;
+          }
+          kDims.push_back(dimIdx.value());
+        }
+        encoding = MatmulKAttr::get(ctx, rewriter.getDenseI32ArrayAttr(kDims));
```

Comment:
There is a builder for `ArrayRef<int32_t>` inputs: 

https://github.com/iree-org/iree/blob/77243062f1638f20eb832772af53f214f5b6370d/compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.td#L238-L240